### DISCUSSION TITLE
Backup Docmost

### DISCUSSION THREAD

**Discussion by proxforge** [docmost/docmost#152] [Ideas]
* admin option to download a fullbackup of docmost data
* admin option to download all spaces with all subpages 

currently I dont know howto backup docmost so in case my server dies, I can recover and do not loose any data.

**Comment by psychoberries** [comment]
Hi, did you figure out how to backup?
I am trying to figure out the same thing, and for me this is quite a problem, because I need to trust that I can restore the content, if anything happens.
I am sure it can be done, but I tried dumping the DB via docker exec pg_dump, but the restore does not seem to produce any effect.

**Comment by sy772** [comment]
hi i am also looking this solution are there any guide or trick.

**Comment by proxforge** [comment]
I did some deep dive into the structure, but I did not succeed. We just need a way to export all pages (the single page export is done via api calls) so I guess implementing this would be easy for the docmost developer. At the moment Im doing a backup of the vm, that hosts this container.

**Reply by adrian-dumitrescu-tfs** [reply]
For what it's worth, I am now doing backups of the volumes


```
#!/bin/bash

TIMESTAMP="$(date -u +%Y%m%d_%H%M%S)"

docker run --rm -v docmost_db_data:/db_data -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_db_data.tar /db_data
docker run --rm -v docmost_redis_data:/redis_data -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_redis_data.tar /redis_data
docker run --rm -v docmost_docmost:/docmost -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_docmost.tar /docmost

mkdir backup/${TIMESTAMP}
cp backup/backup_*.tar backup/${TIMESTAMP}
```

and restore

```
#!/bin/bash

if [ -z "$1" ]; then
  TIMESTAMP="$1"
else
  TIMESTAMP="."
fi

docker run --rm -v docmost_db_data:/db_data -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_db_data.tar
docker run --rm -v docmost_redis_data:/redis_data -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_redis_data.tar
docker run --rm -v docmost_docmost:/docmost -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_docmost.tar
```

**Comment by horusofoz** [comment]
Adding to this as it's mostly related and not different enough to start a new topic.

Is it possible to configure a scheduled backup/export to markdown or similar so as to have access to your documentation if your DocMost instance goes down?

**Reply by bolavefasfas** [reply]
idk though it depends on how the pages are stored in db or locally if in db then you need a script to export them or api if its there if not in db if its doing locally then you can simplily backup the local path.

best idea would be take zfs snapshots of the container dir 😂

**Reply by motusira** [reply]
the pages are stored in the postgresql database, and you can make a dump, but I haven't figured out how to save the attachments and then restore everything or run it on another server yet

**Comment by roleforgaming** [comment]
Curious if there's been any headway on this?

**Comment by Philipinho** [comment]
I do not think backups belongs to the core software.
However, I understand the need for a backup solution and guide.

I will soon be working on the backup documentation page.

**Reply by roleforgaming** [reply]
Thank you! I don't mind there not being a native backup integration, the main issue is that it seems like backing up via direct copy or db dumps does not work because it does not restore properly (well, at all). Thanks for all the work you've put in thus far! I'm loving it, just would be really sad to lose my database and it not be recoverable.

**Reply by ghost** [reply]
I have tried out Docmost several times and I was unable to figure out backups, having this, even if it's just a guide to do it properly would be the last step towards switching from Outline.

**Comment by 5uy4n9** [comment]
**backup database**
`docker exec -t docmost-db-container-name pg_dumpall -c -U docmost | gzip > /path/to/backups/docmost_all_$(date +%Y-%m-%d"_"%H_%M_%S).sql.gz`

**backup files**
`docker run --rm --volumes-from docmost-wiki-container-name -v /path/to/backups:/backup ubuntu tar czf /backup/docmost_file_$(date +"%Y-%m-%d_%H_%M_%S").tar.gz -C /app/data/storage/ .`

**Comment by adrian-dumitrescu-tfs** [comment]
For what it's worth, I am now doing backups of the volumes


```
#!/bin/bash

TIMESTAMP="$(date -u +%Y%m%d_%H%M%S)"

docker run --rm -v docmost_db_data:/db_data -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_db_data.tar /db_data
docker run --rm -v docmost_redis_data:/redis_data -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_redis_data.tar /redis_data
docker run --rm -v docmost_docmost:/docmost -v $(pwd)/backup:/backup ubuntu tar cvf /backup/backup_docmost.tar /docmost

mkdir backup/${TIMESTAMP}
cp backup/backup_*.tar backup/${TIMESTAMP}
```

and restore

```
#!/bin/bash

if [ -z "$1" ]; then
  TIMESTAMP="$1"
else
  TIMESTAMP="."
fi

docker run --rm -v docmost_db_data:/db_data -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_db_data.tar
docker run --rm -v docmost_redis_data:/redis_data -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_redis_data.tar
docker run --rm -v docmost_docmost:/docmost -v $(pwd)/backup:/backup ubuntu tar xvf /backup/${TIMESTAMP}/backup_docmost.tar
```

**Reply by Just-Some-Guy** [reply]
This worked for me.  

I was able to backup and move my site to another server using these scripts and had no problems.

Thanks for the help.

**Comment by changkeke** [comment]
docmost非常棒，我个人在win11上使用docmost，做了个简单的定时提醒备份，以下是我个人的实践仅供参考：
# 一、环境： 
- Win11 64位，docker-desktop，WSL2 integration Ubuntu
```bash
$ uname -a 
Linux omenckk 5.15.167.4-microsoft-standard-WSL2 #1 SMP Tue Nov 5 00:21:55 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -a 
No LSB modules are available. Distributor ID: Ubuntu Description:    Ubuntu 24.04.1 LTS Release:        24.04 Codename:       noble
```
- docmost在我win系统下，使用docker compose直接启动：
```bash
D:\docmost>ls -ll
total 16
-rwxrwx---+ 1 ckk None 1738 Mar  6 10:14 docker-compose.yml
-rwxrwx---+ 1 ckk None 1763 Mar 13 11:49 export-volumes.bat
-rwxrwx---+ 1 ckk None 4016 Mar 13 11:33 import-volumes.bat
-rwxrwx---+ 1 ckk None 1661 Mar 13 12:14 notify-backup.ps1
```

- 我的备份路径是移动硬盘，位置：E:\docker-volumes-export111

- 目标是通过windows右下角每日定时提醒我备份三个存储卷，我可以根据当前的工作情况选择立即执行或取消（后续手动去执行备份脚本）：docmost_db_data、docmost_docmost、docmost_redis_data

# 二、备份与还原：
## 手动导出：
> 脚本功能：立即停止容器、导出卷、启动容器
- 【手动导出】可以双击 `export-volumes.bat` 脚本自动备份到移动硬盘。 
- 注意修改导出目录即可： set EXPORT_DIR=E:\docker-volumes-export111
- 注意修改镜像源：DOCKER_MIRROR

## 添加到win定时提醒：
> 上面的问题可以优化一下，添加到定时任务每天下午6点windows右下角弹窗提醒我。
这时候，我提供了一个powershell的通知脚本`notify-backup.ps1`，把它添加到windows的计划任务。
效果：每天下午6点windows右下角弹窗提醒我，可以选择立即备份或者取消备份。（为什么要做成这样？因为导出操作需要停止容器这可能会导致我正在编写Excalidraw还没保存就丢失了。做成这样以后，我可以先保存我当前的工作内容，然后再去点击立即备份触发备份。）

- 注意修改 `notify-backup.ps1` 中调用导出脚本的路径：D:\docmost\export-volumes.bat

- 然后设置windows计划任务：

打开任务计划程序 按 Win + R 输入 taskschd.msc 点击确定

在右侧操作面板中点击"创建任务"

在"常规"选项卡中： 名称：输入 "Docmost存储卷备份提醒" 选择"使用最高权限运行" 配置：选择 "Windows 10"

4.在"触发器"选项卡： 点击"新建" 设置你想要的备份时间（比如每天下午5点）

5.在"操作"选项卡： 点击"新建" 程序或脚本：C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe 添加参数：-ExecutionPolicy Bypass -File "D:\docmost\notify-backup.ps1" 起始于：脚本所在的文件夹路径

6.在"条件"选项卡： 建议取消勾选"只有在计算机使用交流电源时才启动此任务" 建议取消勾选"只有在空闲时才启动此任务"

7.在"设置"选项卡： 勾选"允许按需运行任务" "如果此任务已经运行，以下规则适用"选择"请勿启动新实例"

注意事项： 确保填写 ps1 文件的完整路径 确保运行账户有足够的权限 建议先手动运行测试一下命令是否正常工作（powershell.exe -ExecutionPolicy Bypass -File "D:\docmost\notify-backup.ps1"） 可以在任务计划程序的日志中查看运行记录

## 手动导入或恢复存储卷：
> 当我需要还原我之前的备份存储卷时执行。比如换了个电脑。
注意：
新环境时，导入的存储卷需要docker-compose up -d命令去创建，不要在导入脚本里加命令去创建存储卷以避免不必要的警告信息。
旧环境还原时，导入存储卷的原理是覆盖掉已有的存储卷内容，这是危险的，我加了很多提示防止手滑。
导入是个不常用的操作，推荐不用此脚本去导入，建议用desktop去导入避免必要的麻烦。我写这个的目的是玩也是经过了多种情况的测试的。

- 【手动导入】如果换电脑了，或者需要恢复卷，双击执行` import-volumes.bat `即可。 
- 注意修改导入目录：set IMPORT_DIR=E:\docker-volumes-export111
- 注意修改镜像源：DOCKER_MIRROR
- 为了防止误覆盖，脚本本身就有y/n确认，不过由于脚本几乎不用，为了更加安全起见，在脚本开始加了10次提示。

## 脚本文件：
export-volumes.bat:
```bash
@echo off
chcp 65001
setlocal enabledelayedexpansion

:: 设置导出目录
set EXPORT_DIR=E:\docker-volumes-export111
:: 设置docker镜像源
set DOCKER_MIRROR=docker.hlmirror.com/
:: 指定临时容器
set TEMP_CONTAINER=%DOCKER_MIRROR%alpine:latest

:: 创建导出目录（如果不存在）
if not exist "%EXPORT_DIR%" mkdir "%EXPORT_DIR%"

:: 列出所有卷以便确认
echo 列出所有docmost Docker卷...
docker volume ls | grep docmost

:: 下载临时容器
echo 下载临时容器...
docker pull %TEMP_CONTAINER%

:: 停止运行中的容器
echo 停止 Docker 容器...
docker-compose down
echo 导出 docmost 数据卷...
docker run --rm -v docmost_docmost:/source -v %EXPORT_DIR%:/dest %TEMP_CONTAINER% sh -c "ls -la /source && tar czf /dest/docmost_docmost.tar.gz -C /source ."

echo 导出 db_data 数据卷...
docker run --rm -v docmost_db_data:/source -v %EXPORT_DIR%:/dest %TEMP_CONTAINER% sh -c "ls -la /source && tar czf /dest/docmost_db_data.tar.gz -C /source ."

echo 导出 redis_data 数据卷...
docker run --rm -v docmost_redis_data:/source -v %EXPORT_DIR%:/dest %TEMP_CONTAINER% sh -c "ls -la /source && tar czf /dest/docmost_redis_data.tar.gz -C /source ."

:: 重新启动容器
echo 重新启动 Docker 容器...
docker-compose up -d

echo 数据卷导出完成！文件保存在: %EXPORT_DIR%

:: 自动打开导出目录
explorer "%EXPORT_DIR%"

:: 删除 pause，替换为完成通知
powershell -Command "Add-Type -AssemblyName System.Windows.Forms; $notify = New-Object System.Windows.Forms.NotifyIcon; $notify.Icon = [System.Drawing.SystemIcons]::Information; $notify.Visible = $true; $notify.ShowBalloonTip(0, '备份完成', '数据卷已成功导出到 %EXPORT_DIR%', 'Info')"
```

notify-backup.ps1:
```bash
# Create WPF window to receive notification responses
Add-Type -AssemblyName System.Windows.Forms
Add-Type -AssemblyName System.Drawing

# Create notification icon
$notify = New-Object System.Windows.Forms.NotifyIcon
$notify.Icon = [System.Drawing.SystemIcons]::Information
$notify.Visible = $true

# Create context menu
$contextMenu = New-Object System.Windows.Forms.ContextMenuStrip

# Backup now option
$menuItemNow = New-Object System.Windows.Forms.ToolStripMenuItem
$menuItemNow.Text = "Backup Now"
$menuItemNow.Add_Click({
    $notify.Dispose()
    Start-Process -FilePath "D:\docmost\export-volumes.bat" -WindowStyle Hidden
    $form.Close()
})

# Cancel option
$menuItemCancel = New-Object System.Windows.Forms.ToolStripMenuItem
$menuItemCancel.Text = "Cancel Backup"
$menuItemCancel.Add_Click({
    $notify.Dispose()
    $form.Close()
})

# Add menu items
$contextMenu.Items.Add($menuItemNow) | Out-Null
$contextMenu.Items.Add("-") | Out-Null
$contextMenu.Items.Add($menuItemCancel) | Out-Null

# Set notification icon context menu
$notify.ContextMenuStrip = $contextMenu

# Show notification
$notify.ShowBalloonTip(
    0,  # Display time (milliseconds)
    "Backup Reminder", 
    "System is ready to perform data backup, please select an action", 
    [System.Windows.Forms.ToolTipIcon]::Info
)

# Create a hidden form to keep the script running
$form = New-Object System.Windows.Forms.Form
$form.WindowState = [System.Windows.Forms.FormWindowState]::Minimized
$form.ShowInTaskbar = $false
$form.Add_Closing({ $notify.Dispose() })

# Run application
[System.Windows.Forms.Application]::Run($form) 
```

import-volumes.bat:
```bash
@echo off
chcp 65001
setlocal enabledelayedexpansion

set TOTAL_WARNINGS=10
set COUNT=1

:warning_loop
echo.
echo [%COUNT%/%TOTAL_WARNINGS%] 导入操作将会覆盖您现有的存储卷数据！请仔细确认您现在的行为！
pause
set /a COUNT+=1
if %COUNT% leq %TOTAL_WARNINGS% goto warning_loop

:: 添加初始检查
echo Starting script...
echo Checking Docker service...
docker info > nul 2>&1
if errorlevel 1 (
    echo ERROR: Docker service is not running!
    goto :end
)

:: 检查卷是否存在
echo Checking if volumes exist...
docker volume ls | findstr "docmost_docmost" > nul
set DOCMOST_EXISTS=%errorlevel%
docker volume ls | findstr "docmost_db_data" > nul
set DB_EXISTS=%errorlevel%
docker volume ls | findstr "docmost_redis_data" > nul
set REDIS_EXISTS=%errorlevel%

:: 如果任何一个卷不存在，则需要启动容器创建卷（保证卷是docker compose创建的！）
if %DOCMOST_EXISTS% neq 0 (
    goto start_containers
) else if %DB_EXISTS% neq 0 (
    goto start_containers
) else if %REDIS_EXISTS% neq 0 (
    goto start_containers
) else (
    echo All volumes already exist, skipping container restart.
    goto continue_import
)

:start_containers
echo One or more volumes do not exist. Starting containers to create them...
echo Restarting Docker containers...
docker-compose down
docker-compose up -d

:: 等待容器启动
echo Waiting for containers to start...
docker-compose ps

:continue_import
:: 设置导入目录（与导出目录相同）
set IMPORT_DIR=E:\docker-volumes-export111
echo Import directory is: %IMPORT_DIR%

:: 设置docker镜像源
set DOCKER_MIRROR=docker.hlmirror.com/
echo Docker mirror is: %DOCKER_MIRROR%

:: 指定临时容器
set TEMP_CONTAINER=%DOCKER_MIRROR%alpine:latest
echo Temporary container is: %TEMP_CONTAINER%

:: 检查导入目录是否存在
if not exist "%IMPORT_DIR%" (
    echo ERROR: Import directory %IMPORT_DIR% does not exist!
    goto :end
)
echo Import directory exists.

:: 检查备份文件是否存在
if not exist "%IMPORT_DIR%\docmost_docmost.tar.gz" (
    echo ERROR: docmost_docmost.tar.gz backup file does not exist!
    goto :end
)
echo docmost_docmost.tar.gz backup file exists.
if not exist "%IMPORT_DIR%\docmost_db_data.tar.gz" (
    echo ERROR: docmost_db_data.tar.gz backup file does not exist!
    goto :end
)
echo docmost_db_data.tar.gz backup file exists.
if not exist "%IMPORT_DIR%\docmost_redis_data.tar.gz" (
    echo ERROR: docmost_redis_data.tar.gz backup file does not exist!
    goto :end
)
echo docmost_redis_data.tar.gz backup file exists.

:: 提示用户是否要覆盖卷
echo WARNING: One or more volumes already exist. Continuing will overwrite existing data!
choice /C YN /M "Do you want to continue"
if errorlevel 2 (
    echo Operation cancelled.
    goto :end
) else (
    echo Proceeding with import...
)

:: 继续执行导入操作
echo Starting import process...

:: 下载临时容器
echo Downloading temporary container...
docker pull %TEMP_CONTAINER%

:: 停止运行中的容器
echo Stopping Docker containers...
docker-compose down

:: 导入 docmost 数据卷
echo Importing docmost volume...
docker run --rm -v docmost_docmost:/dest -v %IMPORT_DIR%:/source %TEMP_CONTAINER% sh -c "rm -rf /dest/* && tar xzf /source/docmost_docmost.tar.gz -C /dest"

:: 导入 db_data 数据卷
echo Importing db_data volume...
docker run --rm -v docmost_db_data:/dest -v %IMPORT_DIR%:/source %TEMP_CONTAINER% sh -c "rm -rf /dest/* && tar xzf /source/docmost_db_data.tar.gz -C /dest"

:: 导入 redis_data 数据卷
echo Importing redis_data volume...
docker run --rm -v docmost_redis_data:/dest -v %IMPORT_DIR%:/source %TEMP_CONTAINER% sh -c "rm -rf /dest/* && tar xzf /source/docmost_redis_data.tar.gz -C /dest"

:: 重新启动容器
echo Restarting Docker containers...
docker-compose up -d

echo Import completed!

:end
:: 暂停显示结果
pause 
```

**Comment by githubkusi** [comment]
Here's my backup/restore solution with [Borg](https://www.borgbackup.org/). Borg takes care of file deduplication, which is an advantage over zipping the whole docker volume for each backup if you have large files attached to Docmost

### Prerequisites
* Install Borg with fuse support for mounting borg repos

        apt install borgbackup python3-pyfuse3

* Adapt docker container & volume names & backup dir in the scripts below
* Add the backup script to /etc/cron.daily


### Backup

        #!/bin/bash

        DB_CONTAINER_NAME=docmost-db-1
        DOCMOST_CONTAINER_NAME=docmost-docmost-1
        DOCMOST_VOLUME=docmost_docmost
        
        # Contains the Borg repo        
        BACKUP_DIR=/backup/docmost

        # Contains the sql dump and the files from the volume
        DATA_DIR=/tmp/docmost-data
                
        # create folder which will contain all files of the docmost volume and the sql file
        if [ -d "$DATA_DIR" ]; then
                rm -r $DATA_DIR
        fi
        mkdir -p $DATA_DIR/volume
                
        # dump database
        docker exec -t $DB_CONTAINER_NAME pg_dumpall -c -U docmost > $DATA_DIR/$DB_CONTAINER_NAME.sql
        
        # copy files from docker volume to tmp folder
        docker run --rm \
            -v "$DOCMOST_VOLUME:/docmost" \
            -v "$DATA_DIR/volume:/backup" \
            alpine sh -c "cp -a /docmost/. /backup/"
        
        # Need to cd into tmp folder otherwise the folder hierarchy becomes part of the backup
        cd $DATA_DIR
        borg create --progress $BACKUP_DIR::docmost-{now} .


### Restore

        #!/bin/bash

        # You first need to mount your desired backup into a tmp folder ($DATA_DIR)
        # E.g borg mount /backup/docmost::docmost-2025-07-13T21:05:30 tmp        

        # Restore overrides the current state
        echo "WARNING: this is a destructive operation, press any key to continue" && read

        DB_CONTAINER_NAME=docmost-db-1
        DOCMOST_CONTAINER_NAME=docmost-docmost-1
        DOCMOST_VOLUME=docmost_docmost
   
        DATA_DIR=$PWD/tmp
        
        docker compose down
        docker compose create
        docker compose start db
        
        # wait for the db container to be started
        sleep 5
        
        # database
        cat $DATA_DIR/$DB_CONTAINER_NAME.sql | docker exec -i "$DB_CONTAINER_NAME" psql -U docmost -d postgres
        
        # files
        docker run --rm \
            -v "$DATA_DIR/volume":/source_local \
            -v "$DOCMOST_VOLUME":/target_volume \
            alpine sh -c "cp -a /source_local/. /target_volume/"
                
        # Check output of compose up for errors. If all works out, call docker compose up -d
        docker compose up
